### PR TITLE
nodePackages.pnpm: 8.15.5 -> 9.1.0

### PR DESCRIPTION
**Pull request by binarious**
## Description of changes

- https://github.com/pnpm/pnpm/releases/tag/v8.15.6
- https://github.com/pnpm/pnpm/releases/tag/v9.0.0
- https://github.com/pnpm/pnpm/releases/tag/v9.0.1
- https://github.com/pnpm/pnpm/releases/tag/v9.0.2
- https://github.com/pnpm/pnpm/releases/tag/v9.0.3
- https://github.com/pnpm/pnpm/releases/tag/v9.0.4
- https://github.com/pnpm/pnpm/releases/tag/v9.0.5
- https://github.com/pnpm/pnpm/releases/tag/v9.0.6
- https://github.com/pnpm/pnpm/releases/tag/v9.1.0

Conflicts with https://github.com/NixOS/nixpkgs/pull/301509 which updates pnpm to the older version 8.15.6.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by binarious** [comment]
@ofborg build nodePackages.pnpm

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 305026` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>geph.gui</li>
    <li>kiwitalk</li>
    <li>pot</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>youtube-music</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>surrealist</li>
  </ul>
</details>

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
We cannot merge this as is, as it breaks 8 out of 9 applications using pnpm. Mostly with the following error message 

```
 ERR_PNPM_BAD_PM_VERSION  This project is configured to use v8.10.2 of pnpm. Your current pnpm is v9.0.2
```

**Comment by wereHamster** [comment]
Do you think it's worth it having both v8 and v9 major versions in nixpkgs? I think it may take a while for those (broken) packages to get updated to v9.

**Comment by SuperSandro2000** [comment]
If we can just ignore the version mismatch and things still build, i am all for that.

**Comment by mikroskeem** [comment]
pnpm 9.0.6 is now available, should update this PR to that?

**Comment by meierw** [comment]
For those that may be waiting on this update.

You have an alternative of using `corepack`, and control your `pnpm` version from the `packageManager` property in `package.json`.

Minimal example using `flake.nix`
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
    flake-utils.url = "github:numtide/flake-utils";
  };

  outputs = { self, nixpkgs, flake-utils, ... }: flake-utils.lib.eachDefaultSystem (
    system:
    let
      pkgs = import nixpkgs { inherit system; };
    in
    {
      devShells.default = pkgs.mkShell {
        buildInputs = [
          pkgs.nodejs-slim_20
          pkgs.corepack_20
        ];
      };
    }
  );
}
```

And then in in your `nix develop` shell.

```shell
corepack use pnpm@latest
```

More info here: https://nodejs.org/api/corepack.html

**Comment by binarious** [comment]
> pnpm 9.0.6 is now available, should update this PR to that?

Updated to 9.1.0, but since we seem to be blocked by depending packages, this game could continue for quite some time.

> Do you think it's worth it having both v8 and v9 major versions in nixpkgs? I think it may take a while for those (broken) packages to get updated to v9.

Sounds like a solution to the situation, but I don't know how to do it. This is my first contribution.

@ofborg build nodePackages.pnpm

File: pkgs/applications/audio/youtube-music/default.nix
Comment (2xsaiko) [inline pkgs/applications/audio/youtube-music/default.nix:54]:
```suggestion
      x86_64-darwin = "sha256-3O/YDZNQ7mTvDF6hRKPekQifOjUBLc4pYYn1WJdtPP4=";
      aarch64-darwin = "sha256-NwTIUxTmB637Ibr6Zuwu3ah3fFtZqfc9usf0I+CAORU=";
```

File: pkgs/by-name/ki/kiwitalk/package.nix
Comment (2xsaiko) [inline pkgs/by-name/ki/kiwitalk/package.nix:73]:
```suggestion
      x86_64-darwin = "sha256-bXtfd41doSi9EV+7swlwqUl1N/aAL2fM4CBHStCP+lI=";
      aarch64-darwin = "sha256-VQg/2daHKSJpdN0KmtxcqExua6yfHaeK+srIJ12eXjA=";
```

**Review by dblsaiko — CHANGES_REQUESTED** [review]
There you go!

**Comment by binarious** [comment]
Updated the hashes. We'd need someone to get the hash of `i686-linux` for `kiwitalk`: https://github.com/NixOS/nixpkgs/pull/305026/files#diff-184ec01f30ccd9c7d88e1e48571d131d7f1d61ba2f14bf7a33adaf1e4c113f1eR71

Who could help?

**Comment by getchoo** [comment]
architecture specific hashes shouldn't be needed if the `--force` flag is passed to `pnpm install` in the deps derivation. i would do that instead to avoid future work

**Comment by binarious** [comment]
@getchoo Thanks, didn't know.

Here we go:
Result of `nixpkgs-review pr 305026` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>geph.gui</li>
    <li>kiwitalk</li>
    <li>pot</li>
    <li>surrealist</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>youtube-music</li>
  </ul>
</details>

**Comment by Scrumplex** [comment]
> architecture specific hashes shouldn't be needed if the `--force` flag is passed to `pnpm install` in the deps derivation. i would do that instead to avoid future work

This is not the full story. Like in https://github.com/NixOS/nixpkgs/pull/290715, you also need to set the following config option for the `pnpmDeps`/`pnpm-deps` derivation:

```sh
# Some packages produce platform dependent outputs. We do not want to cache those in the global store
pnpm config set side-effects-cache false
```

**Comment by Scrumplex** [comment]
Result of `nixpkgs-review pr 305026` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>kiwitalk</li>
    <li>vikunja</li>
    <li>youtube-music</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>geph.gui</li>
    <li>pot</li>
    <li>surrealist</li>
    <li>vesktop</li>
  </ul>
</details>

**Comment by binarious** [comment]
@Scrumplex Will add side-effect-cache config option. What build failures did you get?

**Comment by Scrumplex** [comment]
> What build failures did you get?

Hash mismatches for the packages mentioned above. Without disabling the side effects cache, it can produce different outputs on different systems

**Comment by binarious** [comment]
With that added the build of vikunja fails:
> ERR_PNPM_PATCH_FAILED  Could not apply patch /build/source/frontend/patches/flexsearch@0.7.31.patch to /build/source/frontend/node_modules/.pnpm/flexsearch@0.7.31_patch_hash=bfn3sngfuhktmdj7jgl3ejl35y/node_modules/flexsearch

**Comment by binarious** [comment]
@Scrumplex Could you please retry? I might have had the wrong hash of vikunja and a cache of it in my store. Maybe we're lucky and vikunja does not produce different hashes on different environments.

**Comment by kolaente** [comment]
Vikunja maintainer here, not sure why the patch did not apply correctly. Vikunja uses Cypress in development which downloads platform-specific binaries so I'd imagine the hash is different for each platform.
That being said, Cypress is only used in tests so it should be safe to disable downloading these binaries for a more reproucible output: https://docs.cypress.io/guides/references/advanced-installation#Skipping-installation

FWITW, the official Vikunja builds in our CI are already using pnpm 9 and that seems to work fine.

**Comment by binarious** [comment]
@kolaente Thanks for letting us know. `--ignore-scripts` should already skip `postinstall` and thus the downloads, shouldn't it?

**Comment by kolaente** [comment]
> Thanks for letting us know. --ignore-scripts should already skip postinstall and thus the downloads, shouldn't it?

@binarious That sounds like it should be the case.

**Comment by natsukium** [comment]
Should we wait to merge until branch off?

**Comment by gepbird** [comment]
9.1.1 was recently released https://github.com/pnpm/pnpm/releases/tag/v9.1.1

**Comment by marsam** [comment]
Another alternative could be to move pnpm as a top-level package
```nix
# pkgs/by-name/pn/pnpm/package.nix
{ lib, stdenvNoCC, fetchurl, nodejs-slim, withNode ? false, testers }:

stdenvNoCC.mkDerivation (finalAttrs: {
  pname = "pnpm";
  version = "9.1.1";

  src = fetchurl {
    url = "https://registry.npmjs.org/pnpm/-/pnpm-${finalAttrs.version}.tgz";
    hash = "sha256-lVHoA9y3oYOf31QWFTqEQGDHvOATIYzoI0EFMlBKwQs=";
  };

  buildInputs = lib.optionals withNode [ nodejs-slim ];

  installPhase = ''
    runHook preInstall

    mkdir -p $out/{bin,libexec}
    cp -R . $out/libexec/pnpm
    ln -s $out/libexec/pnpm/bin/pnpm.cjs $out/bin/pnpm
    ln -s $out/libexec/pnpm/bin/pnpx.cjs $out/bin/pnpx

    runHook postInstall
  '';

  passthru = {
    tests.version = lib.optionalAttrs withNode (testers.testVersion {
      package = finalAttrs.finalPackage;
    });
  };

  meta = with lib; {
    description = "Fast, disk space efficient package manager";
    homepage = "https://pnpm.io/";
    license = licenses.mit;
    platforms = platforms.all;
    maintainers = [ ];
  };
})
```

So packages that require pnpm 8.x, can just override it

**Comment by binarious** [comment]
@Scrumplex So should we close this PR and wait for yours to be done or should I update pnpm again, resolve the conflicts and we merge it until it's replaced with your solution?

**Comment by Scrumplex** [comment]
> So should we close this PR and wait for yours to be done or should I update pnpm again, resolve the conflicts and we merge it until it's replaced with your solution?

I don't really have an ETA for the PR. Perhaps you can adopt my top-level pnpm package into this PR. That way we can have `pnpm_8` and `pnpm_9` which should help reduce breakages caused by the new major version

**Comment by SuperSandro2000** [comment]
Now that #290715 got merged, please rebase

**Comment by linsui** [comment]
We shuold update the pnpm version for those pnpm packages and remove the override in all-packages.

**Comment by binarious** [comment]
> Now that #290715 got merged, please rebase.
> 
> Maybe we don't even need this PR at this point, then please remove pnpm from nodePackages.

I'm going to close this PR since its intention was to update pnpm to version 9 which was resolved via https://github.com/NixOS/nixpkgs/pull/290715 if I understand it correctly. Anything else should probably go into a fresh PR.

**Pull request closed without merging**
